### ISSUE TITLE
Algorithm Benchmark in Orion

### ISSUE THREAD

**Issue by donglinjy** [Epistimio/orion#415]
**Why need benchmark**

Algorithms benchmark could be useful in some scenarios:
- Understand the capabilities of difference algorithms so people can choose the proper algorithm in different user scenarios.
- It could be a useful tool to help developer to evaluate the correctness and performance of new search algorithms.

**How it can be done**

To run benchmark, we need define the goal (Assessment Scenarios) of the benchmark and tasks (whose hyper-parameters will be optimized) that will be used.

Some possible Assessment Scenarios can be:
- Time to result
- Average performance
- Search space dimension size (the number of hyperparameters to search)
- Search space dimension Type (Rang vs Categorical)
- Parallel Execution advantage (does more worker means better performance?)
- Search Algorithm parameters sensitivity (easy to use)

And Tasks to run can be divided into 3 categories:
- Synthetic functions, e.g. Branin, rosenbrock
- Emulator/Simulator, e.g. PROFET, HPOlib
- Real verification tasks, e.g. MNIST, cifar10 

From high level view, we can run benchmark with below mechanism. 
<img width="550" alt="image" src="https://user-images.githubusercontent.com/10850941/85941309-bdbea200-b954-11ea-9de0-843ce147bb60.png">

And this can be the rough code sketch.
```
def create_benchmark():
    """
    - create a benchmark object based on user input
    - run the benchmark in backend(thread) and return the benchmark object
    - or run the benchmark in sequence and ask user to call the benchmark process()
    :return:
    """
    pass


class Benchmark():
    def __init__(self):
        """
        - build assess and task pairs
        """
        pass

    def process(self):
        """
        - for each assess-task, build Assess object and run assess execute
        - support multiple thread based a configured thread_num ?
        :return:
        """
        pass

    def status(self):
        """
        - get progress/details of each assess-task
        :return:
        """
        pass

    def analysis(self):
        """
        - display the result of each assess
        :return:
        """
        pass

    def register(self):
        """
        register benchmark object into db
        :return:
        """
        pass

class BaseAssess():

    def __init__(self, task):
        """
        - build assess object
        - build task object (from db[existing], or from config[new])
        """
        pass

    def execute(self):
        """
        - run the tasks
        - there may be needs to run the task multiple times (such as when assess average performance)
        :return:
        """
        pass

    def status(self):
        """
        - get the overall status of the assess, like how many tasks to run and the status of each task(experiment)
        :return:
        """
        pass

    def result(self):
        """
        -  json format of the result
        :return:
        """
        pass

    def display(self):
        """
        - define the visual charts of the assess, based on the task performance output
        :return:
        """
        pass

    def register(self):
        """
        register assess object into db
        :return:
        """
        pass


class BaseTask():

    # assessments that the particular task supports
    assessments = ['timetoresult', 'averagerank']

    def __init__(self, algorithm):
        """
        - build orion experiment
        """
        pass

    def run(self):
        """
        - run the orion experiment
        :return:
        """
        pass

    def status(self):
        """
        - status of the orion experiment
        """
        pass

    def performance(self):
        """
        - formatted the experiment result for the particular assess
        :return:
        """
        pass

```

**Comment by donglinjy** [comment]
Initial work for this is on https://github.com/donglinjy/orion/tree/feature/benchmark. 
The rough plan would be:
1. Have a rough benchmark flow to support benchmark with some typical synthetic functions 
2. Support some typical real tasks, such as classification, regression, etc.
3. Support to run assessment against tasks generated by simulator or emulator.
4. Refine the execution flow for better performance, such as supporting parallel execution.

**Comment by donglinjy** [comment]
example notebook: https://github.com/donglinjy/orion/blob/feature/benchmark/examples/benchmark/benchmark_get_start.ipynb

**Comment by bouthilx** [comment]
Done in #513
